### PR TITLE
Jakob/structure

### PR DESCRIPTION
**Pull request by jmmaljaars**
Restructuring the code, and adding some l2 projection functionality

### CODE REVIEW

File: src/advect/advect.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+#pragma once
+
+#include "../Particles.h"
+
+namespace leopart
+{
+namespace advect
+{
+// Advection related code should go into leopart::advect namespace
+class Advect
```
Comment (chrisrichardson) [inline src/advect/advect.h:15]: I am not sure having an `Advect` class is necessary.

File: src/io/leopart_io.h [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// Copyright: (c) 2020 Chris Richardson and Jakob Maljaars
+// This file is part of LEoPart-X, a particle-in-cell package for DOLFIN-X
+// License: GNU Lesser GPL version 3 or any later version
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+#pragma once
+
+#include "../Particles.h"
```
Comment (chrisrichardson) [inline src/io/leopart_io.h:8]: Should avoid having "../" in the code. Better to use `<Particles.h>` and make sure that the path is in the include path.

File: src/advect/advect.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
+// Copyright: (c) 2020 Jakob Maljaars, Chris Richardson and Nate Sime
+// This file is part of LEoPart-X, a particle-in-cell package for DOLFIN-X
+// License: GNU Lesser GPL version 3 or any later version
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+#pragma once
+
+#include "../Particles.h"
```
Comment (chrisrichardson) [inline src/advect/advect.h:8]: avoid "../"

File: src/project/PDEproject.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace leopart
+{
+namespace project
+{
+
+class PDEProjection
+{
+public:
+  PDEProjection(const Particles& pax){};
```
Comment (chrisrichardson) [inline src/project/PDEproject.h:18]: Why does this need to be a class?

File: src/wrapper.cpp [hunk 5]
```diff
@@ -42,25 +54,36 @@ PYBIND11_MODULE(pyleopart, m)
...
-  m.def("random_tet", &generation::random_reference_tetrahedron);
-  m.def("random_tri", &generation::random_reference_triangle);
-  m.def("mesh_fill", &generation::mesh_fill);
+  m.def("random_tet", &random_reference_tetrahedron);
+  m.def("random_tri", &random_reference_triangle);
+  m.def("mesh_fill", &mesh_fill);
 
-  // Transfer functions
-  m.def("get_particle_contributions", &transfer::get_particle_contributions);
+  // Transfer functions (probably shouldn't be interfaced)
```
Comment (chrisrichardson) [inline src/wrapper.cpp:73]: Maybe OK for testing in pytest. They might not be in the "public" interface.

File: src/io/leopart_io.h (hunk 2 above)
Comment (jmmaljaars) [reply src/io/leopart_io.h]: Deleted

File: src/advect/advect.h (hunk 3 above)
Comment (jmmaljaars) [reply src/advect/advect.h]: See above

File: src/wrapper.cpp (hunk 5 above)
Comment (jmmaljaars) [reply src/wrapper.cpp:73]: Not sure how we can decide "after" the wrapper whether something should be publicly interfaced or not?

Comment (chrisrichardson) [reply src/wrapper.cpp:73]: One option would be to give them a name starting with "_" which is public, but showing that it is not really for general use.

**Review by chrisrichardson — APPROVED** [review]

File: src/io/leopart_io.h (hunk 2 above)
Comment (chrisrichardson) [reply src/io/leopart_io.h]: looks good

**Pull request merged**